src/Tasks/MapSourceRoots.cs

```diff
@@ -0,0 +1,117 @@
+                    {
+                        string nestedRoot = root.GetMetadata(Names.NestedRoot);
+                        if (string.IsNullOrEmpty(nestedRoot))
+                        {
+                            if (topLevelMappedPaths.ContainsKey(root.ItemSpec))
+                            {
+                                Log.LogErrorFromResources("MapSourceRoots.ContainsDuplicate", Names.SourceRoot, root.ItemSpec);
+                                success = false;
+                            }
+                            else
+                            {
+                                var mappedPath = "/_" + (i == 0 ? "" : i.ToString()) + "/";
+                                topLevelMappedPaths.Add(root.ItemSpec, mappedPath);
+                                root.SetMetadata(Names.MappedPath, mappedPath);
+                                i++;
+                            }
+                        }
+                    }
+                }
+            }
+
+            string EndWithSlash(string path)
+                => (path[path.Length - 1] == '/') ? path : path + '/';
+
+            // assign mapped paths to process source control roots first:
+            SetTopLevelMappedPaths(sourceControl: true);
+
+            // then assign mapped paths to other source control roots:
+            SetTopLevelMappedPaths(sourceControl: false);
+
+            // finally, calculate mapped paths of nested roots:
+            foreach (var root in SourceRoots)
+            {
+                string nestedRoot = root.GetMetadata(Names.NestedRoot);
+                if (!string.IsNullOrEmpty(nestedRoot))
+                {
+                    string containingRoot = root.GetMetadata(Names.ContainingRoot);
+
+                    // The value of ContainingRoot metadata is a file path that is compared with ItemSpec values of SourceRoot items.
+                    // Since the paths in ItemSpec have backslashes replaced with slashes on non-Windows platforms we need to do the same for ContainingRoot.
+                    if (containingRoot != null && topLevelMappedPaths.TryGetValue(FileUtilities.FixFilePath(containingRoot), out var mappedTopLevelPath))
+                    {
+                        Debug.Assert(mappedTopLevelPath.EndsWith("/", StringComparison.Ordinal));
+                        root.SetMetadata(Names.MappedPath, mappedTopLevelPath + EndWithSlash(nestedRoot.Replace('\\', '/')));
+                    }
+                    else
+                    {
+                        Log.LogErrorFromResources("MapSourceRoots.ValueOfNotFoundInItems", Names.SourceRoot + "." + Names.ContainingRoot, Names.SourceRoot, containingRoot);
+                        success = false;
+                    }
+                }
+            }
+
+            if (success)
+            {
+                MappedSourceRoots = SourceRoots;
+            }
+
+            return success;
```

Comment:
In our experience, the best practice for returning from `Task.Execute` is

```c#
return !Log.HasLoggedErrors;
```

That helps avoid the confusing situation where you fail a task but have no error in the log about why.

Reply:
OK.